### ISSUE TITLE
File.ReadAllTextAsync not behaving asynchronously when called on inaccessible network files

### ISSUE THREAD

**Issue by cjwainwright** [dotnet/corefx#27724]
When accessing a file on a network share that for some reason cannot be accessed (maybe not existent), `File.ReadAllTextAsync` appears to block, rather than running asynchronously, until the file request fails. This can be reproduced with the following console app:

```csharp
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Threading.Tasks;

namespace ReadAllTextAsync
{
    class Program
    {
        public static void Main(string[] args)
        {
            Console.WriteLine("Beginning");

            int count = 20;
            var files = Enumerable.Range(0, count).Select(i => $@"\\server_{Guid.NewGuid():N}\file.txt").ToList();
            var tasks = new List<Task>();

            foreach (string file in files)
            {
                tasks.Add(_ReadFile(file));
            }

            Console.WriteLine("Waiting...");

            Task.WhenAll(tasks).Wait();

            Console.WriteLine("Ending");
        }

        private static async Task _ReadFile(string file)
        {
            try
            {
                // await Task.Delay(1); // Uncomment this line allows calls to execute in parallel
                string text = await File.ReadAllTextAsync(file);
            }
            catch
            {
                Console.WriteLine($"Error reading file: {file}");
            }
        }
    }
}
```
The result of running the above code is a regular stream of (the expected) "Error reading file" messages but they come in slowly one after the other.

When uncommenting the `await Task.Delay(1)` line, the messages all come in nearly simulataneously. I would have expected this behaviour without this line.

I believe this is to do with the behaviour of the `new FileStream` object in https://github.com/dotnet/corefx/blob/release/2.0.0/src/System.IO.FileSystem/src/System/IO/File.cs

```csharp
        private static StreamReader AsyncStreamReader(string path, Encoding encoding)
        {
            FileStream stream = new FileStream(
                path, FileMode.Open, FileAccess.Read, FileShare.Read, DefaultBufferSize,
                FileOptions.Asynchronous | FileOptions.SequentialScan);

            return new StreamReader(stream, encoding, detectEncodingFromByteOrderMarks: true);
        }
```
which is called before any code is awaited.

**Comment by danmosemsft** [comment]
cc @JonHanna who added these, in case he is interested.

**Comment by JonHanna** [comment]
Yeah, the problem being in the ctor would make sense.

Ideally we'd have a `ObtainFileStreamAsync()` we could use (does something like that already exist?). Failing that we could `await Task.Yield()` just before we get `stream`, or perhaps do so if it's a UNC path and not otherwise.

**Comment by svick** [comment]
@JonHanna 

>  Failing that we could await `Task.Yield()` just before we get `stream`

I don't think that would be a good workaround, since it would still block the calling synchronization context. This matters especially in GUI apps, where it would still block the UI thread.

Doing something like `var stream = await Task.Run(() => new FileStream(…));` should be better, since it's guaranteed to block a threadpool thread.

**Comment by JonHanna** [comment]
My idea about about only doing this for a UNC path isn't a good one, as mapped drives hide UNC and there can be other slow-to-access files.

Is there anything better than `var stream = await Task.Run(() => new FileStream(…));` that anyone knows of? Otherwise I'll apply that?

**Comment by stephentoub** [comment]
> Otherwise I'll apply that?

Any caller of this method could do the same thing if they need to, but doing it in the implementation means it's being forced on to everyone.

**Comment by JonHanna** [comment]
Not really, as obtaining the `FileStream` is within the method, so there's no way of either opting in or opting out of this.

If there isn't a better way, maybe that's the real problem here, and this should wait on such a better way being implemented first.

**Comment by stephentoub** [comment]
> Not really, as obtaining the FileStream is within the method, so there's no way of either opting in or opting out of this.

Sure there is:
```C#
Task<string> task = Task.Run(() => File.ReadAllTextAsync(path));
```

> maybe that's the real problem here

Right, Win32 doesn't provide any API that enables performing the open asynchronously.

**Comment by danmosemsft** [comment]
-> Future as this seems to haev a workaround. cc @pjanotti
